### PR TITLE
Draft: Adds team model and first unit test.

### LINKED ISSUE #1: Team Generation Core
## Feature Description
The purpose of the team generation algorithm is to shuffle teams based on their game skill to get random teams which in average have an equal skill based on a specific game.
The team generation algorithm consists of several steps:

* Check how many teams can be generated based on the amount of player and team size
* Sort player by their game skill
* Put player ascending into teams alternating between first and last team
* Calculate team skill of each team
* Refine teams bi swapping players between team with highest and lowest skill (repeat team_amount/2 times)

If team size is not divisible by amount of players (without remainder) then create potential replacement player lists for teams that are not full.

### PR DESCRIPTION
**Pull request by tgrafena**
Closes #1 

Merge request ready to be merged.

### CODE REVIEW

File: src/models/TeamGenerator.ts [hunk 1]
```diff
@@ -197,20 +198,57 @@ export default class TeamGenerator {
...
         //Step 4: If any swap possibilities, pick one randomly and return true, if not return false
+        if (potentialSwapPairs.length > 0){
+            let randomIndex: number = Math.floor(Math.random() * (potentialSwapPairs.length - 1));
+            let [swapPlayerHigh, swapPlayerLow] = potentialSwapPairs[randomIndex];
+            higherTeam.removePlayer(swapPlayerHigh);
+            lowerTeam.removePlayer(swapPlayerLow);
+            higherTeam.addPlayer(swapPlayerLow);
+            lowerTeam.addPlayer(swapPlayerHigh);
+            return true;
+        }else{
```
Comment (benitsch) [inline src/models/TeamGenerator.ts:305]: You can get rid of this if-else by just write if (potentialSwapPairs.length <= 0) {return false}. Then it covers only the exceptional case and the other code is executed normally (without indentation).

File: src/models/TeamGenerator.ts [hunk 2]
```diff
@@ -178,6 +178,20 @@ export default class TeamGenerator {
...
+    /**
+     * This function is a protected subroutine call which takes 2 full teams and 1 game as input
+     * and tries to swap 1 player pair if applicable to decrease the skill diff of the given teams
+     * as much as possible. If several options with the same result are available then a random
+     * pair out of the options is chosen.
+     * 
+     * NOTE!: this function does no param check as it is only a subroutine and expects the given teams
+     * to be full and the game skill is given for all players for the passed game.
+     * 
+     * @param team1 The first team to be balanced.
```
Comment (benitsch) [inline src/models/TeamGenerator.ts:248]: Not sure if you want to add the type in the JSDoc as well? e.g. @param {Team} team1

File: src/models/Team.ts [hunk 3]
```diff
@@ -27,20 +32,53 @@ export default class Team {
...
+        return this._playerList.length + this._substitutionPlayerList.length;
+    }
+
+    get isFull(): boolean {
+        return this.currentSize == this.targetSize;
     }
 
     addPlayer(player: Player): boolean {
-        if (!this._playerList.includes(player) && this.currentSize < this.targetSize){
+        if (!this.isFull && !this.isPlayerInTeam(player)){
```
Comment (benitsch) [inline src/models/Team.ts:43]: The whole function can be simplified by:
return (!this.isFull && !this.isPlayerInTeam(player)) ? (this._playerList.push(player), true) : false;

File: src/models/Team.ts [hunk 4]
```diff
@@ -27,20 +32,53 @@ export default class Team {
...
             this._playerList.push(player);
             return true;
         }else {
             return false;
         }
     }
 
+    addSubstitutionPlayer(player: Player): boolean {
+        if (!this.isFull && !this.isPlayerInTeam(player)){
```
Comment (benitsch) [inline src/models/Team.ts:52]: Same here. Shorter/simpler see addPlayer func review.

File: src/models/TeamGenerator.ts [hunk 5]
```diff
@@ -75,47 +75,102 @@ export default class TeamGenerator {
...
+
+        return fullTeams;
+    }
+
+    /**
+     * 
+     * @param players 
+     * @param teamSize 
+     * @param game 
+     * @returns
```
Comment (benitsch) [inline src/models/TeamGenerator.ts:115]: Do you want to add the return type? e.g. @returns {(GeneratorErrorCode|undefined)}

File: src/models/TeamGenerator.ts [hunk 6]
```diff
@@ -75,47 +75,102 @@ export default class TeamGenerator {
...
+        return orderedPlayerArray;
+    }
 
-        
-        //Step 3: calculate the amount of teams creatable out of the player list and create them(players.length % teamSize may be > 0)
-        // amount of teams that can be filled up with players
+    /**
+     * 
+     * @param orderedPlayerArray 
+     * @param teamSize
```
Comment (benitsch) [inline src/models/TeamGenerator.ts:158]: Missing param "game" in JSDoc

File: src/models/Player.ts [hunk 7]
```diff
@@ -58,6 +73,27 @@ export default class Player {
         this._gameSkills.push(skill);
     }
 
+    isSkillAssessedForGame(game: Game): boolean{
+        for (const gameSkill of this._gameSkills){
+            if(gameSkill.game === game){
+                if (gameSkill.skillLevel > 0){
+                    return true;
+                }
+                break;
```
Comment (benitsch) [inline src/models/Player.ts:82]: Are you sure a "break" is correct here? Or do you want to "continue" here to continue search within the gameSkills?

File: src/models/TeamGenerator.ts (hunk 1 above)
Comment (tgrafena) [reply src/models/TeamGenerator.ts:305]: True but what could be gained here with the proposed change? Do you think it improves readability? Also "length" can never be 0 right so "==" would be more reasonable.

File: src/models/TeamGenerator.ts (hunk 2 above)
Comment (tgrafena) [reply src/models/TeamGenerator.ts:248]: Ah nice, I did not know that (or not anymore). Thanks for the hint. Will do.

File: src/models/Team.ts (hunk 3 above)
Comment (tgrafena) [reply src/models/Team.ts:43]: Done.

File: src/models/Team.ts (hunk 4 above)
Comment (tgrafena) [reply src/models/Team.ts]: Done.

File: src/models/TeamGenerator.ts (hunk 5 above)
Comment (tgrafena) [reply src/models/TeamGenerator.ts:115]: Hm... at least in VSCode I do see the return type only in the signature when the function description is displayed. @return does not show the type. I copy pasted your suggestion as it is.

File: src/models/TeamGenerator.ts (hunk 6 above)
Comment (tgrafena) [reply src/models/TeamGenerator.ts:158]: Added. Please resolve at will.

File: src/models/TeamGenerator.ts (hunk 2 above)
Comment (tgrafena) [reply src/models/TeamGenerator.ts:286]: I tried to add this but somehow in VS code this doesn't show in the documentation.

File: src/models/Player.ts (hunk 7 above)
Comment (tgrafena) [reply src/models/Player.ts:82]: Yes "break" is intentional as the game has been found so we do not need to look further. There are two cases where we want to return false. 

1)  The game has not been found
2)  The game has not been assessed

Maybe it is more clear if i refactor it to the following:

```
for (const gameSkill of this._gameSkills){
        if(gameSkill.game === game){
            return (gameSkill.skillLevel > 0) ? true : false;
        }
}
return false;
```

File: src/models/TeamGenerator.ts (hunk 1 above)
Comment (tgrafena) [reply src/models/TeamGenerator.ts:350]: After we talked I see where the intention comes from (handle all error  cases at the beginning) but this is an error case that happens at the end and cannot be handled at the beginning.  Shall I still apply your proposed changes? Just let me know and then we can resolve this and merge. ;)

**Pull request merged**